### PR TITLE
archive: store the override xattr with the inode type

### PR DESCRIPTION
**Pull request by giuseppe**
Closes: https://github.com/containers/storage/issues/2174

### CODE REVIEW

**Comment by giuseppe** [comment]
related fuse-overlayfs change: https://github.com/containers/fuse-overlayfs/pull/434

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -660,7 +660,17 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		mask = *forceMask
 	}
 
-	switch hdr.Typeflag {
+	typeFlag := hdr.Typeflag
+
+	if forceMask != nil {
+		// If we have a forceMask, force the real type to either be a directory,
+		// or a regular file.
+		if typeFlag != tar.TypeDir {
```
Comment (mtrmac) [inline pkg/archive/archive.go:668]: What about symlinks? This turns them into regular files, and the destination seems to be lost.

File: pkg/chunked/storage_linux.go [hunk 2]
```diff
@@ -1487,7 +1487,7 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 	for i := range mergedEntries {
 		r := &mergedEntries[i]
 		if options.ForceMask != nil {
-			value := idtools.FormatContainersOverrideXattr(r.UID, r.GID, int(r.Mode))
+			value := idtools.FormatContainersOverrideXattrDevice(r.UID, r.GID, int(r.Mode), int(r.Devmajor), int(r.Devminor))
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:1490]: Doesn’t this need to handle `r.Type`?

And there is a `SetContainersOverrideXattr` call in this file that seems to need updating.

File: pkg/idtools/idtools.go [hunk 3]
```diff
@@ -369,54 +370,124 @@ func checkChownErr(err error, name string, uid, gid int) error {
...
+			if err != nil {
+				return stat, err
+			}
+		} else if strings.HasPrefix(typ, "char") {
+			stat.Mode |= os.ModeCharDevice
+			stat.Major, stat.Minor, err = parseDevice(typ)
+			if err != nil {
+				return stat, err
+			}
+		}
```
Comment (mtrmac) [inline pkg/idtools/idtools.go:466]: Else fail?

File: pkg/idtools/idtools.go [hunk 4]
```diff
@@ -436,7 +507,7 @@ func SafeChown(name string, uid, gid int) error {
 				}
 			}
 		}
-		value := Stat{IDPair{uid, gid}, mode}
+		value := Stat{IDPair{uid, gid}, mode, 0, 0}
```
Comment (mtrmac) [inline pkg/idtools/idtools.go:510]: Does this need to handle file types?

File: pkg/idtools/idtools.go [hunk 5]
```diff
@@ -464,7 +535,7 @@ func SafeLchown(name string, uid, gid int) error {
...
```
Comment (mtrmac) [inline pkg/idtools/idtools.go:538]: Does this need to handle file types?

File: pkg/idtools/idtools.go [hunk 6]
```diff
@@ -436,7 +507,7 @@ func SafeChown(name string, uid, gid int) error {
 				}
```
Comment (mtrmac) [inline pkg/idtools/idtools.go:507]: The parser above expects exactly 3 fields. Same in `SafeLchown`.

**Review by mtrmac — COMMENTED** [review]
ACK conceptually, but there seems even more to be done.

- symbolic links are now not handled consistently.
- `check.go` interprets `force_mask`; it needs to be taught about this
- Doesn’t the code in `overlay.Driver.create` around creating `diff` need to also set `os.ModeDir` now?
- The parser in `pkg/archive.remapIDs` seems to need updating (preferably consolidate all parsers to one function)

(It gets confusing… if only the OS had separate types for “permission bits” and “the full mode including a `ModeType`.)

**Comment by giuseppe** [comment]
pushed a new version, and verified manually both partial pulls and with regular pulls

File: pkg/idtools/idtools_test.go [hunk 7]
```diff
@@ -273,3 +276,155 @@ func TestIsContiguous(t *testing.T) {
...
+		minor    int
+		expected string
+	}{
+		{"regular file", 1000, 1001, 0o644, 0, 0, "1000:1001:0644:file"},
+		{"directory", 1000, 1001, 0o755 | int(os.ModeDir), 0, 0, "1000:1001:0755:dir"},
+		{"symlink", 1000, 1001, 0o777 | int(os.ModeSymlink), 0, 0, "1000:1001:0777:symlink"},
+		{"pipe", 1000, 1001, 0o666 | int(os.ModeNamedPipe), 0, 0, "1000:1001:0666:pipe"},
+		{"socket", 1000, 1001, 0o700 | int(os.ModeSocket), 0, 0, "1000:1001:0700:socket"},
+		{"block device", 1000, 1001, 0o600 | int(os.ModeDevice), 8, 0, "1000:1001:0600:block-8-0"},
+		{"char device", 1000, 1001, 0o600 | int(os.ModeDevice|os.ModeCharDevice), 1, 3, "1000:1001:0600:char-1-3"},
```
Comment (mtrmac) [inline pkg/idtools/idtools_test.go:389]: Also test permission handling - at least `ISUID`/`ISGID`/`ISVTX`, or possibly each of the 12 bits.

**Review by mtrmac — COMMENTED** [review]
ACK to the symlink semantics change.

Note also the outstanding items in the top-level review comment in https://github.com/containers/storage/pull/2183#pullrequestreview-2473801817 .

File: pkg/idtools/idtools_test.go [hunk 8]
```diff
@@ -273,3 +278,180 @@ func TestIsContiguous(t *testing.T) {
...
+		{"directory", 1001, 1012, fs.FileMode(0o755) | os.ModeDir, 0, 0, "1001:1012:0755:dir"},
+		{"symlink", 1002, 1013, fs.FileMode(0o777) | os.ModeSymlink, 0, 0, "1002:1013:0777:symlink"},
+		{"pipe", 1003, 1014, fs.FileMode(0o666) | os.ModeNamedPipe, 0, 0, "1003:1014:0666:pipe"},
+		{"socket", 1004, 1015, fs.FileMode(0o700) | os.ModeSocket, 0, 0, "1004:1015:0700:socket"},
+		{"block device", 1005, 1016, fs.FileMode(0o600) | os.ModeDevice, 8, 0, "1005:1016:0600:block-8-0"},
+		{"char device", 1006, 1017, fs.FileMode(0o600) | os.ModeDevice | os.ModeCharDevice, 1, 3, "1006:1017:0600:char-1-3"},
+
+		{"setuid bit", 1012, 1022, fs.FileMode(0o4755), 0, 0, "1012:1022:4755:file"},
+		{"setgid bit", 1013, 1023, fs.FileMode(0o2755), 0, 0, "1013:1023:2755:file"},
+		{"sticky bit", 1014, 1024, fs.FileMode(0o1777), 0, 0, "1014:1024:1777:file"},
```
Comment (mtrmac) [inline pkg/idtools/idtools_test.go:395]: `07000` are not the right `fs.Mode*` values; AFAICS they have no assigned meaning.

**Review by mtrmac — COMMENTED** [review]
Don’t `SafeChown`/`SafeLchown` still need a bit of an update?

**Comment by giuseppe** [comment]
> Don’t `SafeChown`/`SafeLchown` still need a bit of an update?

I've changed them further. I don't really know how they are used on darwin.  I see the same data is also set on the file itself.

**Comment by mtrmac** [comment]
IIRC the general idea has been to allow unprivileged macOS users to extract volumes, but I’d need to look up the git history to confirm.

File: pkg/idtools/idtools_test.go [hunk 9]
```diff
@@ -273,3 +278,180 @@ func TestIsContiguous(t *testing.T) {
...
+		{"directory", 1001, 1012, fs.FileMode(0o755) | os.ModeDir, 0, 0, "1001:1012:0755:dir"},
+		{"symlink", 1002, 1013, fs.FileMode(0o777) | os.ModeSymlink, 0, 0, "1002:1013:0777:symlink"},
+		{"pipe", 1003, 1014, fs.FileMode(0o666) | os.ModeNamedPipe, 0, 0, "1003:1014:0666:pipe"},
+		{"socket", 1004, 1015, fs.FileMode(0o700) | os.ModeSocket, 0, 0, "1004:1015:0700:socket"},
+		{"block device", 1005, 1016, fs.FileMode(0o600) | os.ModeDevice, 8, 0, "1005:1016:0600:block-8-0"},
+		{"char device", 1006, 1017, fs.FileMode(0o600) | os.ModeDevice | os.ModeCharDevice, 1, 3, "1006:1017:0600:char-1-3"},
+
+		{"setuid bit", 1012, 1022, fs.FileMode(0o4755) | fs.ModeSetuid, 0, 0, "1012:1022:4755:file"},
+		{"setgid bit", 1013, 1023, fs.FileMode(0o2755) | fs.ModeSetgid, 0, 0, "1013:1023:2755:file"},
+		{"sticky bit", 1014, 1024, fs.FileMode(0o1777) | fs.ModeSticky, 0, 0, "1014:1024:1777:file"},
```
Comment (mtrmac) [inline pkg/idtools/idtools_test.go:395]:
```suggestion
		{"setuid bit", 1012, 1022, fs.FileMode(0o755) | fs.ModeSetuid, 0, 0, "1012:1022:4755:file"},
		{"setgid bit", 1013, 1023, fs.FileMode(0o755) | fs.ModeSetgid, 0, 0, "1013:1023:2755:file"},
		{"sticky bit", 1014, 1024, fs.FileMode(0o777) | fs.ModeSticky, 0, 0, "1014:1024:1777:file"},
```
?

**Review by mtrmac — COMMENTED** [review]
LGTM otherwise.

File: pkg/idtools/idtools_test.go (hunk 9 above)
Comment (giuseppe) [reply pkg/idtools/idtools_test.go]: amended

**Pull request merged**